File: src/net/dinkla/raytracer/btdf/BTDF.java
package net.dinkla.raytracer.btdf;

import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.math.Vector3DF;
import net.dinkla.raytracer.samplers.Sampler;

/**
 * Created by IntelliJ IDEA.
 * User: Jörn Dinkla
 * Date: 01.05.2010
 * Time: 09:53:32
 * To change this template use File | Settings | File Templates.
 */
public abstract class BTDF<C extends Color> {

    Sampler sampler;

    public class Sample {
        public C color;
        public Vector3DF wt;
    }

    abstract public C f(Shade sr, Vector3DF wo, Vector3DF wi);

    abstract public Sample sampleF(Shade sr, Vector3DF wo);

    abstract public C rho(Shade sr, Vector3DF wo);

    abstract public boolean isTir(Shade sr);
}

File: src/main/kotlin/net/dinkla/raytracer/btdf/BTDF.kt
package net.dinkla.raytracer.btdf

import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.math.Vector3D
import net.dinkla.raytracer.samplers.Sampler

abstract class BTDF {

    class Sample(
            val color: Color = Color.WHITE,
            val wt: Vector3D = Vector3D.ZERO
    )

    abstract fun f(sr: Shade, wo: Vector3D, wi: Vector3D): Color
    abstract fun sampleF(sr: Shade, wo: Vector3D): Sample
    abstract fun rho(sr: Shade, wo: Vector3D): Color
    abstract fun isTir(sr: Shade): Boolean
}
